ctl/string.cc

```diff
@@ -64,29 +75,37 @@ string::string(const char* s, size_t size) noexcept
 const char*
 string::c_str() const noexcept
 {
-    if (!n)
-        return "";
-    if (n >= c)
+    if (size() >= capacity())
         __builtin_trap();
-    if (p[n])
+    if (data()[size()])
         __builtin_trap();
-    return p;
+    return data();
 }
 
 void
 string::reserve(size_t c2) noexcept
 {
     char* p2;
-    if (c2 < n)
-        c2 = n;
+    if (c2 < size())
+        c2 = size();
     if (ckd_add(&c2, c2, 15))
         __builtin_trap();
     c2 &= -16;
-    if (!(p2 = (char*)realloc(p, c2)))
-        __builtin_trap();
+    if (c2 <= __::sso_max)
+        return;
+    if (!isbig()) {
+        if (!(p2 = (char *)malloc(c2)))
+            __builtin_trap();
+        __builtin_memcpy(p2, data(), size() + 1);
```

Comment:
`__builtin_memcpy()` would make sense for a lightweight header, but in the `.cc` file you can do anything.